sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_arg.asciidoc

```diff
@@ -0,0 +1,812 @@
+:cpp: C++
+:dpcpp: DPC++
+
+== Notice
+
+[%hardbreaks]
+Copyright (c) 2021-2022 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+https://github.com/intel/llvm/issues
+
+== Contributors
+
+Joe Garvey, Intel +
+Abhishek Tiwari, Intel
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 4 and
+the following extensions:
+
+- link:../experimental/sycl_ext_oneapi_properties.asciidoc[sycl_ext_oneapi_properties]
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback. Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state. The specification itself may also change in
+incompatible ways before it is finalized. *Shipping software products should not
+rely on APIs defined in this specification.*
+
+== Overview
+This extension introduces a wrapper class that provides a mechanism to attach
+compile-time constant information to a pointer in a manner that allows the
+compiler to reliably maintain and analyze the information.
+
+== Overview
+
+The purpose of this document is to clearly describe and specify
+`annotated_arg` and related concepts, types, and mechanisms, and to give
+examples and context for their usage.
+
+[NOTE]
+====
+In this document, we use the shortened form `annotated_arg` to refer to the
+proposed `sycl::ext::oneapi::annotated_arg` class.
+====
+
+The purpose of this document is to clearly describe and specify `annotated_arg`
+and related concepts, types, and mechanisms, and to give examples and context
+for their usage.
+
+== Specification
+=== Feature Test Macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_ANNOTATED_ARG` to one of the values defined
+in the table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's APIs the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version.  Base features are supported.
+|===
+
+=== Introduction
+
+The `annotated_arg` class enables users to attach compile-time constant
+information to the kernel arguments with properties.
+
+The example below shows a use case with some properties `PropA` and `PropB`.
+
+.Toy Example
+[source,c++]
+----
+using sycl::ext::oneapi::experimental;
+struct MyKernel {
+  annotated_arg<int, properties<PropA>> a;
+  annotated_arg<int, properties<PropB>> b;
+  int *c;
```

Comment:
The fact that this is a regular pointer really stands out, and made me think about the difference between `annotated_ptr<int>` and `annotated_arg<int*>`. The specification talks later about exposing `operator*` and `operator[]`, but it might not be obvious to all readers that using a pointer here is valid.

It might be interesting to declare this as an `annotated_arg<int*>`, then add a paragraph explaining what that means. 
Based on the rest of the extension, my understanding is we want to say something like:

> `annotated_arg<T*>` applies properties to the storage location associated with a pointer variable. These properties do not interact with the object the pointer points to, and have no effect on the behavior of dereferencing the pointer. The [sycl_ext_oneapi_annotated_ptr](../proposed/sycl_ext_oneapi_annotated_ptr.asciidoc) extension provides a closely related `annotated_ptr<T>` class for use-cases where properties affect usage of the pointer.

Reply:
Thanks for your review John, yes your understanding is correct. I like your suggestion a lot, I will make the update.

Reply:
Updated.